### PR TITLE
Enabling build with clang compiler for fuzzing

### PR DESCRIPTION
**Pull request by robertwoj-microsoft**
## Description

Fix some issues reported by clang compiler.

## Checklist

- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-osconfig/blob/main/CONTRIBUTING.md).
- [x] All unit tests are passing.
- [x] I have merged the latest `main` branch prior to this PR submission.
- [x] I submitted this PR against the `main` branch.

### CODE REVIEW

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -20,11 +20,11 @@ option(BUILD_MODULETEST "Build the moduletest tool" ON)
 option(BUILD_SAMPLES "Build samples" OFF)
 option(COVERAGE "Enable code coverage" OFF)
 
-if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU" OR CMAKE_C_COMPILER_ID STREQUAL "Clang")
     add_compile_options("-Wno-psabi;-fPIC")
    
     # Special Ubuntu 14/GNU 4.8 C compiler build configuration for the Universal NRP adapter
-    if (CMAKE_C_COMPILER_VERSION VERSION_GREATER "4.7" AND CMAKE_C_COMPILER_VERSION VERSION_LESS "4.9")
+    if (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER "4.7" AND CMAKE_C_COMPILER_VERSION VERSION_LESS "4.9")
```
Comment (MariusNi) [inline src/CMakeLists.txt:27]: >if (CMAKE_C_COMPILER_ID STREQUAL "GNU" AND CMAKE_C_COMPILER_VERSION VERSION_GREATER "4.7" AND CMAKE_C_COMPILER_VERSION VERSION_LESS "4.9") [](http://example.com/codeflow?start=4&length=138)

Do we have a clang compiler version equivalent to gcc 4.8 and that runs on Ubuntu 14?

File: src/cmake/templates/version.h.in [hunk 2]
```diff
@@ -5,6 +5,6 @@
 #ifndef VERSION_H
 #define VERSION_H
 
-#define OSCONFIG_VERSION R""""(@OsConfigVersionString@)""""
+#define OSCONFIG_VERSION "@OsConfigVersionString@"
```
Comment (MariusNi) [inline src/cmake/templates/version.h.in:8]: >"@OsConfigVersionString@" [](http://example.com/codeflow?start=25&length=25)

Why doing this change?

File: src/modules/CMakeLists.txt [hunk 3]
```diff
@@ -1,7 +1,13 @@
 # Copyright (c) Microsoft Corporation. All rights reserved.
 # Licensed under the MIT License.
 
-add_compile_options("-Wall;-Wextra;-Wunused;-Werror;-Wformat;-Wformat-security;-Wno-unused-result;-Wunused-const-variable=2")
```
Comment (MariusNi) [inline src/modules/CMakeLists.txt:4]: >-Wunused-const-variable=2" [](http://example.com/codeflow?start=98&length=26)

Why did you remove this?

File: src/cmake/templates/version.h.in (hunk 2 above)
Comment (MariusNi) [reply src/cmake/templates/version.h.in:8]: Not sure. If the new format builds and format correctly the version string in both gcc and clang builds, we're good. Can you please check and confirm?

File: src/modules/CMakeLists.txt [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
 # Licensed under the MIT License.
 
-add_compile_options("-Wall;-Wextra;-Wunused;-Werror;-Wformat;-Wformat-security;-Wno-unused-result;-Wunused-const-variable=2")
+add_compile_options("-Wall;-Wextra;-Wunused;-Werror;-Wformat;-Wformat-security;-Wno-unused-result")
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
+    add_compile_options("-Wunused-const-variable=2")
+endif()
+if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    add_compile_options("-Wunused-variable")
+endif()
```
Comment (MariusNi) [inline src/modules/CMakeLists.txt:11]: Can you try using an 'elseif' here to avoid two separate sets of if and endif?

Like this:

if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
    add_compile_options("-Wunused-const-variable=2")
elseif (CMAKE_C_COMPILER_ID STREQUAL "Clang")
    add_compile_options("-Wunused-variable")
endif()

File: src/platform/CMakeLists.txt [hunk 5]
```diff
@@ -1,7 +1,13 @@
...
 # Licensed under the MIT License.
 
-add_compile_options("-Wall;-Wextra;-Wunused;-Werror;-Wformat;-Wformat-security;-Wno-unused-result;-Wunused-const-variable=2")
+add_compile_options("-Wall;-Wextra;-Wunused;-Werror;-Wformat;-Wformat-security;-Wno-unused-result")
+if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
+    add_compile_options("-Wunused-const-variable=2")
+endif()
+if (CMAKE_C_COMPILER_ID STREQUAL "Clang")
+    add_compile_options("-Wunused-variable")
+endif()
```
Comment (MariusNi) [inline src/platform/CMakeLists.txt:10]: Same as other comment. Does this work? This:

if (CMAKE_C_COMPILER_ID STREQUAL "GNU")
    add_compile_options("-Wunused-const-variable=2")
elseif (CMAKE_C_COMPILER_ID STREQUAL "Clang")
    add_compile_options("-Wunused-variable")
endif()

**Review by MariusNi — APPROVED** [review]
:shipit:

**Review by robertschaedler3 — APPROVED** [review]
